post-til-private/src/main/java/no/difi/meldingsutveksling/dpi/json/JWT2XmlSoapDpiReceiptConverter.java

```diff
@@ -0,0 +1,167 @@
+package no.difi.meldingsutveksling.dpi.json;
+
+import com.nimbusds.jose.Payload;
+import lombok.SneakyThrows;
+import no.difi.begrep.sdp.schema_v10.*;
+import no.difi.meldingsutveksling.domain.Iso6523;
+import no.difi.meldingsutveksling.domain.sbdh.ObjectFactory;
+import no.difi.meldingsutveksling.domain.sbdh.*;
+import no.difi.meldingsutveksling.dpi.client.domain.messagetypes.*;
+import no.difi.meldingsutveksling.dpi.client.internal.UnpackJWT;
+import no.difi.meldingsutveksling.dpi.client.internal.UnpackStandardBusinessDocument;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.w3.xmldsig.*;
+
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import javax.xml.stream.XMLOutputFactory;
+import javax.xml.stream.XMLStreamWriter;
+import java.io.ByteArrayOutputStream;
+import java.nio.charset.StandardCharsets;
+import java.time.OffsetDateTime;
+import java.time.ZonedDateTime;
+
+public class JWT2XmlSoapDpiReceiptConverter implements DpiReceiptConverter {
+
+    private final JAXBContext context;
+    private final XMLOutputFactory xmlOutputFactory;
+    private final ObjectFactory objectFactory;
+    private final UnpackJWT unpackJWT;
+    private final UnpackStandardBusinessDocument unpackStandardBusinessDocument;
+
+    public JWT2XmlSoapDpiReceiptConverter(UnpackJWT unpackJWT, UnpackStandardBusinessDocument unpackStandardBusinessDocument) throws JAXBException {
+        this.context = JAXBContext.newInstance(StandardBusinessDocument.class, SDPKvittering.class, SDPFeil.class);
+        this.xmlOutputFactory = XMLOutputFactory.newFactory();
+        this.objectFactory = new ObjectFactory();
+        this.unpackJWT = unpackJWT;
+        this.unpackStandardBusinessDocument = unpackStandardBusinessDocument;
+    }
+
+    @Override
+    @SneakyThrows
+    public String apply(String jwt) {
+        Payload payload = unpackJWT.getPayload(jwt);
+        StandardBusinessDocument sbd = unpackStandardBusinessDocument.unpackStandardBusinessDocument(payload);
+        StandardBusinessDocumentHeader sbdh = sbd.getStandardBusinessDocumentHeader();
+
+        DpiMessageType messageType = DpiMessageType.fromType(sbd.getType());
+
+        sbdh.getSender().stream().map(Partner::getIdentifier).forEach(p -> p.setAuthority(getAuthority(p)));
+        sbdh.getReceiver().stream().map(Partner::getIdentifier).forEach(p -> p.setAuthority(getAuthority(p)));
+        sbdh.getDocumentIdentification()
+                .setStandard("urn:no:difi:sdp:1.0")
+                .setTypeVersion("1.0")
+                .setType(getType(messageType));
+        sbd.getScope(ScopeType.CONVERSATION_ID).ifPresent(p -> p.setIdentifier("urn:no:difi:sdp:1.0"));
+        sbd.setAny(messageType == DpiMessageType.FEIL ? getSdpFeil(sbd) : getSdpKvittering(sbd, messageType));
+
+        Marshaller marshaller = getMarshaller();
```

Comment:
Denne kan vel eksistere på klassenivå, så slipper den å lages for hver kvittering?

Reply:
Nei, marshalleren er ikke thread safe. Den er uansett ikke tung å opprette: https://stackoverflow.com/questions/7400422/jaxb-creating-context-and-marshallers-cost